superset-frontend/src/views/datasetList/DatasetModal.tsx

```diff
@@ -0,0 +1,318 @@
+};
+
+interface Props {
+  addDangerToast: (msg: string) => void;
+  addSuccessToast: (msg: string) => void;
+  onChange: any;
+  onDatasourceSave: any;
+  onHide: any;
+  show: boolean;
+}
+
+interface State {
+  datasource: option;
+  datasources: Array<option>;
+  disableSave: boolean;
+  disableSelectSchema: boolean;
+  disableSelectTable: boolean;
+  schema: option;
+  schemaLoading: boolean;
+  schemas: Array<option>;
+  table: option;
+  tableLoading: boolean;
+  tables: Array<option>;
+}
+
+const FieldTitle = styled.p`
+  color: ${({ theme }) => theme.colors.secondary.light2};
+  font-size: ${({ theme }) => theme.typography.sizes.s};
+  margin: 20px 0 10px 0;
+  text-transform: uppercase;
+`;
+
+const StyledIcon = styled(Icon)`
+  margin: auto 10px auto 0;
+`;
+
+class DatasetModal extends React.PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.onDatabaseChange = this.onDatabaseChange.bind(this);
+    this.onSave = this.onSave.bind(this);
+    this.onSchemaChange = this.onSchemaChange.bind(this);
+    this.onTableChange = this.onTableChange.bind(this);
+    this.renderFields = this.renderFields.bind(this);
+    this.renderTitle = this.renderTitle.bind(this);
+  }
+
+  state: State = {
+    datasource: { label: 'Select Datasource', value: 'Select Datasource' },
+    datasources: [],
+    disableSave: true,
+    disableSelectSchema: true,
+    disableSelectTable: true,
+    schema: { label: 'Select Schema', value: 'Select Schema' },
+    schemaLoading: false,
+    schemas: [],
+    table: { label: 'Select Table', value: 'Select Table' },
+    tableLoading: false,
+    tables: [],
+  };
+
+  componentDidMount() {
+    this.fetchDatabase();
+  }
+
+  onDatabaseChange(datasource: option) {
+    this.setState({
+      disableSave: true,
+      disableSelectSchema: true,
+      disableSelectTable: true,
+      schemas: [],
+      tables: [],
+    });
+    this.fetchSchemas(`${datasource.value}`);
+    this.setState({
+      datasource,
+      schema: { label: 'Select Schema', value: 'Select Schema' },
+      table: { label: 'Select Table', value: 'Select Table' },
+    });
+  }
+
+  onSchemaChange(schema: option) {
+    this.setState({ tables: [], disableSelectTable: true, disableSave: true });
+    this.fetchTables(`${schema.value}`);
+    this.setState({ schema });
+  }
+
+  onTableChange(table: option) {
+    const { disableSelectSchema, disableSelectTable } = this.state;
+    const disableSave = disableSelectSchema || disableSelectTable;
+    this.setState({ table, disableSave });
+  }
+
+  onSave() {
+    const { datasource, schema, table } = this.state;
+    const { onHide, addSuccessToast, addDangerToast } = this.props;
+    const data = {
+      database: datasource.value,
+      schema: schema.value,
+      table_name: table.label,
+    };
+
+    SupersetClient.post({
+      endpoint: '/api/v1/dataset/',
+      body: JSON.stringify(data),
+      headers: { 'Content-Type': 'application/json' },
```

Comment:
i'm shocked you need to define this header here, i'd think there'd be a param or flag to set this content type. Maybe take a look into the SupersetClient code and make sure. you might not even need to set it at all!

Reply:
@kristw @suddjian This seems relevant to this discussion: https://github.com/apache-superset/superset-ui/pull/554#discussion_r433432209